packages/core/src/api/message.ts

```diff
@@ -0,0 +1,182 @@
+import { makeURLSearchParams, type RawFile, type REST } from '@discordjs/rest';
+import {
+	Routes,
+	type APIMessage,
+	type APIUser,
+	type RESTGetAPIChannelMessageReactionUsersQuery,
+	type RESTPostAPIChannelMessageJSONBody,
+} from 'discord-api-types/v10';
+
+export class MessagesAPI {
+	public constructor(private readonly rest: REST) {}
+
+	/**
+	 * Sends a message in a channel
+	 *
+	 * @param channelId - The id of the channel to send the message in
+	 * @param message - The options to use when sending the message
+	 */
+	public async send(channelId: string, message: RESTPostAPIChannelMessageJSONBody & { files?: RawFile[] }) {
+		const messageOptions = typeof message === 'string' ? { content: message } : message;
+
+		return (await this.rest.post(Routes.channelMessages(channelId), {
+			files: messageOptions.files,
+			body: messageOptions,
+		})) as APIMessage;
+	}
+
+	/**
+	 * Edits a message
+	 *
+	 * @param channelId - The id of the channel the message is in
+	 * @param messageId - The id of the message to edit
+	 * @param options - The options to use when editing the message
+	 * @returns
+	 */
+	public async edit(channelId: string, messageId: string, options: RESTPostAPIChannelMessageJSONBody | string) {
+		const messageOptions = typeof options === 'string' ? { content: options } : options;
+
+		return (await this.rest.patch(Routes.channelMessage(channelId, messageId), {
+			body: messageOptions,
+		})) as APIMessage;
+	}
+
+	/**
+	 * Deletes a message
+	 *
+	 * @param channelId - The id of the channel the message is in
+	 * @param messageId - The id of the message to delete
+	 */
+	public async delete(channelId: string, messageId: string) {
+		return (await this.rest.delete(Routes.channelMessage(channelId, messageId))) as APIMessage;
+	}
+
+	/**
+	 * Bulk deletes messages
+	 *
+	 * @param channelId - The id of the channel the messages are in
+	 * @param messageIds - The ids of the messages to delete
+	 */
+	public async bulkDelete(channelId: string, messageIds: string[]): Promise<void> {
+		await this.rest.post(Routes.channelBulkDelete(channelId), {
+			body: {
+				messages: messageIds,
+			},
+		});
+	}
+
+	/**
+	 * Replies to a message
+	 *
+	 * @param channelId - The id of the channel the message is in
+	 * @param messageRef - The message to reply to
+	 * @param options - The options to use when replying to the message
+	 */
+	public async reply(channelId: string, messageRef: APIMessage, options: RESTPostAPIChannelMessageJSONBody | string) {
```

Comment:
Maybe the `messageRef` could allow passing just the `id` & `channel_id` as well as a full APIMessage? i.e. `APIMessage | { id: Snowflake; channel_id: Snowflake }`